src/Control/Monad/Trans/Indexed/State.hs

```diff
@@ -43,14 +63,112 @@ evalStateIx m i = fst <$> runStateIx m i
 execStateIx :: Monad m => StateIx i j m x -> i -> m j
 execStateIx m i = snd <$> runStateIx m i
 
-modifyIx :: Applicative m => (i -> j) -> StateIx i j m ()
-modifyIx f = StateIx $ \i -> pure ((), f i)
-
-putIx :: Applicative m => j -> StateIx i j m ()
-putIx j = modifyIx (\ _ -> j)
-
 toStateT :: StateIx i i m x -> StateT i m x
 toStateT (StateIx f) = StateT f
 
 fromStateT :: StateT i m x -> StateIx i i m x
 fromStateT (StateT f) = StateIx f
+
+class
+  ( IxMonadTrans t
+  , forall m s i j. (Monad m, s ~ i, i ~ j) => MonadState s (t i j m)
+  ) => IxMonadTransState t where
+  {-# MINIMAL putIx | stateIx #-}
+  putIx :: Monad m => j -> t i j m ()
+  putIx s = stateIx (\_ -> ((), s))
+  modifyIx :: Monad m => (i -> j) -> t i j m ()
+  modifyIx f = stateIx (\i -> ((), f i))
+  stateIx :: Monad m => (i -> (a,j)) -> t i j m a
+  stateIx f =
+    bindIx (\s -> let ~(a, s') = f s in thenIx (return a) (putIx s')) get
+instance IxMonadTransState StateIx where
+  stateIx f = StateIx (return . f)
+
+newtype CodensityIx t i j m a = CodensityIx
+  { runCodensityIx :: forall b k. (a -> t j k m b) -> t i k m b }
+  deriving Functor
```

Comment:
The indexed codensity monad transformer type which I found in [sessiontypes](https://github.com/Ferdinand-vW/sessiontypes) by @Ferdinand-vW.